system/man-pages/man/man7/x25.7
.\" Copyright 1998, Heiner Eisen
.\" Copyright, the authors of the Linux man-pages project
.\"
.\" SPDX-License-Identifier: Linux-man-pages-1-para
.\"
.TH x25 7 2026-02-08 "Linux man-pages (unreleased)"
.SH NAME
x25
\-
ITU-T X.25 / ISO/IEC\~8208 protocol interface
.SH SYNOPSIS
.nf
.B #include <sys/socket.h>
.B #include <linux/x25.h>
.P
.IB x25_socket " = socket(AF_X25, SOCK_SEQPACKET, 0);"
.fi
.SH DESCRIPTION
X25 sockets provide an interface to the X.25 packet layer protocol.
This allows applications to
communicate over a public X.25 data network as standardized by
International Telecommunication Union's recommendation X.25
(X.25 DTE-DCE mode).
X25 sockets can also be used for communication
without an intermediate X.25 network (X.25 DTE-DTE mode) as described
in ISO/IEC\~8208.
.P
Message boundaries are preserved \[em] a
.BR read (2)
from a socket will
retrieve the same chunk of data as output with the corresponding
.BR write (2)
to the peer socket.
When necessary, the kernel takes care
of segmenting and reassembling long messages by means of
the X.25 M-bit.
There is no hard-coded upper limit for the
message size.
However, reassembling of a long message might fail if
there is a temporary lack of system resources or when other constraints
(such as socket memory or buffer size limits) become effective.
If that
occurs, the X.25 connection will be reset.
.SS Socket addresses
The
.B AF_X25
socket address family uses the
.I struct sockaddr_x25
for representing network addresses as defined in ITU-T
recommendation X.121.
.P
.in +4n
.EX
struct sockaddr_x25 {
    sa_family_t sx25_family;    /* must be AF_X25 */
    x25_address sx25_addr;      /* X.121 Address */
};
.EE
.in
.P
.I sx25_addr
contains a char array
.I x25_addr[]
to be interpreted as a null-terminated string.
.I sx25_addr.x25_addr[]
consists of up to 15 (not counting the terminating null byte) ASCII
characters forming the X.121 address.
Only the decimal digit characters from \[aq]0\[aq] to \[aq]9\[aq] are allowed.
.SS Socket options
The following X.25-specific socket options can be set by using
.BR setsockopt (2)
and read with
.BR getsockopt (2)
with the
.I level
argument set to
.BR SOL_X25 .
.TP
.B X25_QBITINCL
Controls whether the X.25 Q-bit (Qualified Data Bit) is accessible by the
user.
It expects an integer argument.
If set to 0 (default),
the Q-bit is never set for outgoing packets and the Q-bit of incoming
packets is ignored.
If set to 1, an additional first byte is prepended
to each message read from or written to the socket.
For data read from
the socket, a 0 first byte indicates that the Q-bits of the corresponding
incoming data packets were not set.
A first byte with value 1 indicates
that the Q-bit of the corresponding incoming data packets was set.
If the first byte of the data written to the socket is 1, the Q-bit of the
corresponding outgoing data packets will be set.
If the first byte is 0,
the Q-bit will not be set.
.SH VERSIONS
The AF_X25 protocol family is a new feature of Linux 2.2.
.SH BUGS
Plenty, as the X.25 PLP implementation is
.BR CONFIG_EXPERIMENTAL .
.P
This man page is incomplete.
.P
There is no dedicated application programmer's header file yet;
you need to include the kernel header file
.IR <linux/x25.h> .
.B CONFIG_EXPERIMENTAL
might also imply that future versions of the
interface are not binary compatible.
.P
X.25 N-Reset events are not propagated to the user process yet.
Thus,
if a reset occurred, data might be lost without notice.
.SH SEE ALSO
.BR socket (2),
.BR socket (7)
.P
Jonathan Simon Naylor:
\[lq]The Re-Analysis and Re-Implementation of X.25.\[rq]
The URL is
.UR ftp:\://ftp.pspt.fi\:/pub\:/ham\:/linux\:/ax25\:/x25doc.tgz
.UE .


system/net-tools/lib/x25.c
/*
 * lib/x25.c	This file contains an implementation of the "X.25"
 *		support functions for the NET-3 base distribution.
 *
 * Version:	@(#)x25.c	1.00	08/15/98
 *
 * Author:	Stephane Fillod, <sfillod@charybde.gyptis.frmug.org>
 *		based on ax25.c by:
 * 		Fred N. van Kempen, <waltje@uwalt.nl.mugnet.org>
 *		Copyright 1993 MicroWalt Corporation
 *
 *		This program is free software; you can redistribute it
 *		and/or  modify it under  the terms of  the GNU General
 *		Public  License as  published  by  the  Free  Software
 *		Foundation;  either  version 2 of the License, or  (at
 *		your option) any later version.
 */
#include "config.h"

#if HAVE_AFX25 || HAVE_HWX25
#include <sys/types.h>
#include <sys/ioctl.h>
#include <sys/socket.h>
#include <linux/x25.h>
#include <net/if_arp.h>		/* ARPHRD_X25 */
#include <stdlib.h>
#include <stdio.h>
#include <ctype.h>
#include <errno.h>
#include <fcntl.h>
#include <string.h>
#include <termios.h>
#include <unistd.h>
#include "net-support.h"
#include "pathnames.h"
#define  EXTERN
#if 0
#include "net-locale.h"
#endif
#include "intl.h"
#include "util.h"

static char X25_errmsg[128];


extern struct aftype x25_aftype;

/* is in net/x25.h, not in the public header file linux/x25.h. Why?*/
#ifndef X25_ADDR_LEN
#define X25_ADDR_LEN 16
#endif


static const char *
X25_print(const char *ptr)
{
  static char buff[X25_ADDR_LEN+1];

  strncpy(buff, ptr, X25_ADDR_LEN);
  buff[X25_ADDR_LEN] = '\0';
  return(buff);

}


/* Display an X.25 socket address. */
static const char *
X25_sprint(const struct sockaddr_storage *sasp, int numeric)
{
  const struct sockaddr *sap = (const struct sockaddr *)sasp;

  (void)numeric; /* unused */
  if (sap->sa_family == 0xFFFF || sap->sa_family == 0)
    return( _("[NONE SET]"));
  return(X25_print(((struct sockaddr_x25 *)sasp)->sx25_addr.x25_addr));
}


/*
 * return the sigdigits of the address
 */
static int
X25_input(int type, char *bufp, struct sockaddr_storage *sasp)
{
  struct sockaddr *sap = (struct sockaddr *)sasp;
  char *ptr;
  char *p;
  unsigned int sigdigits;

  (void)type; /* unused */
  sap->sa_family = x25_aftype.af;
  ptr = ((struct sockaddr_x25 *)sap)->sx25_addr.x25_addr;


  /* Address the correct length ? */
  if (strlen(bufp)>18) {
        safe_strncpy(X25_errmsg,
                     _("Address can't exceed eighteen digits with sigdigits"),
                     sizeof(X25_errmsg));
#ifdef DEBUG
        fprintf(stderr, "x25_input(%s): %s !\n", bufp, X25_errmsg);
#endif
        errno = EINVAL;
        return(-1);
  }


  if ((p = strchr(bufp, '/')) != NULL) {
        *p = '\0';
        sigdigits = atoi(p + 1);
  } else {
        sigdigits = strlen(bufp);
  }

  if (strlen(bufp) < 1 || strlen(bufp) > 15 || sigdigits > strlen(bufp)) {
        if (p != NULL)
                *p = '/';
        safe_strncpy(X25_errmsg, _("Invalid address"), sizeof(X25_errmsg));
#ifdef DEBUG
        fprintf(stderr, "x25_input(%s): %s !\n", bufp, X25_errmsg);
#endif
        errno = EINVAL;
        return(-1);
  }

  strncpy(ptr, bufp, sigdigits+1);

  /* All done. */
#ifdef DEBUG
  fprintf(stderr, "x25_input(%s)\n", bufp);
#endif

  return sigdigits;
}


/* Display an error message. */
static void
X25_herror(const char *text)
{
  if (text == NULL) fprintf(stderr, "%s\n", X25_errmsg);
    else fprintf(stderr, "%s: %s\n", text, X25_errmsg);
}


static int
X25_hinput(char *bufp, struct sockaddr_storage *sasp)
{
  struct sockaddr *sap = (struct sockaddr *)sasp;
  if (X25_input(0, bufp, sasp) < 0) return(-1);
  sap->sa_family = ARPHRD_X25;
  return(0);
}


struct hwtype x25_hwtype = {
  "x25",	NULL, /*"CCITT X.25",*/		ARPHRD_X25,	X25_ADDR_LEN,
  X25_print,	X25_hinput,	NULL
};

struct aftype x25_aftype =
{
    "x25", NULL, /*"CCITT X.25", */ AF_X25, X25_ADDR_LEN,
    X25_print, X25_sprint, X25_input, X25_herror,
    X25_rprint, X25_rinput, NULL /* getmask */,
    -1,
    "/proc/net/x25"
};


#endif	/* HAVE_xxX25 */
